eZ/Publish/Core/FieldType/RichText/Normalizer/CharacterEntityDefinition.php

```diff
@@ -0,0 +1,147 @@
+<?php
+
+/**
+ * This file is part of the eZ Publish Kernel package.
+ *
+ * @copyright Copyright (C) eZ Systems AS. All rights reserved.
+ * @license For full copyright and license information view LICENSE file distributed with this source code.
+ *
+ * @version //autogentag//
+ */
+namespace eZ\Publish\Core\FieldType\RichText\Normalizer;
+
+use eZ\Publish\Core\FieldType\RichText\Normalizer;
+
+/**
+ * Character entity definition normalizer adds DTD containing character entity definition to
+ * string input that conforms to an XML document with configured document element and default
+ * namespace.
+ *
+ * Note: if input already contains DTD it won't be accepted for normalization.
+ */
+class CharacterEntityDefinition extends Normalizer
+{
+    /**
+     * Holds root element name of the accepted XML format.
+     *
+     * @var string
+     */
+    private $documentElement;
+
+    /**
+     * Holds default namespace name of the accepted XML format.
+     *
+     * @var string
+     */
+    private $namespace;
+
+    /**
+     * Holds a map of character entities with entity names as keys.
+     *
+     * If the mapped value is integer number, it will be handled as a
+     * numeric character reference.
+     *
+     * @var array
+     */
+    private $entityMap;
+
+    /**
+     * Holds computed regular expression pattern for matching and replacement.
+     *
+     * @var string
+     */
+    private $expression;
+
+    /**
+     * Holds computed XML DTD for input normalization.
+     *
+     * @var string
+     */
+    private $definition;
+
+    public function __construct($documentElement, $namespace, array $entityMap)
+    {
+        $this->documentElement = $documentElement;
+        $this->namespace = $namespace;
+        $this->entityMap = $entityMap;
+    }
+
+    /**
+     * Accept if $input looks like XML document, with configured document element
+     * and default namespace, but without DTD.
+     *
+     * @param string $input
+     *
+     * @return bool
+     */
+    public function accept($input)
+    {
+        if (preg_match($this->getExpression(), $input, $matches)) {
+            return true;
+        }
+
+        return false;
+    }
+
+    /**
+     * Normalizes given $input by adding DTD with character entity definition.
+     *
+     * @param string $input
+     *
+     * @return string
+     */
+    public function normalize($input)
+    {
+        if (!empty($this->entityMap)) {
+            return preg_replace(
```

Comment:
> manipulating XML with some regexp is usually a bad idea. I think you can achieve the same by using a custom DOMDocumentType and then create a DOMDocument with that type.

I thought that might provide something for loading XML with character entities but without DTD defining them, which is essentially not well-formed. But I don't see how this can be done.
I know regex is not particularly suitable for XML, but short of implementing custom XML parser in PHP, I'm out of options. AFAIK this is simplest/cheapest way to normalize input.

> Alternatively, normalize could also load a DOM Document with the DTD and return the XML with numeric entities instead of injecting the DTD in the document but I'm unsure on how this can be done to be honest.

I don't see it either :)